core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -4799,6 +4811,77 @@ && isHourKeyword(qc.getAst().token) && qc.getAst().paramCount == 1 && qc.getAst(
         return root;
     }
 
+    private boolean checkIfTranslatingModelIsRedundant(boolean useInnerModel, boolean useGroupByModel, boolean useWindowModel, boolean forceTranslatingModel, boolean checkTranslatingModel, QueryModel translatingModel) {
+        // check if translating model is redundant, e.g.
+        // that it neither chooses between tables nor renames columns
+        boolean translationIsRedundant = (useInnerModel || useGroupByModel || useWindowModel) && !forceTranslatingModel;
+        if (translationIsRedundant && checkTranslatingModel) {
+            for (int i = 0, n = translatingModel.getBottomUpColumns().size(); i < n; i++) {
+                QueryColumn column = translatingModel.getBottomUpColumns().getQuick(i);
+                if (!Chars.equalsIgnoreCase(column.getAst().token, column.getAlias())) {
+                    translationIsRedundant = false;
+                    break;
+                }
+            }
+        }
+        return translationIsRedundant;
+    }
+
+    /**
+     * Checks if an alias appears in the columns and function args of a model.
+     * <p>
+     * Consider a query:<br>
+     * (select x1, sum(x1) from (select x as x1 from y))<br>
+     * If you were to move lift the column from the RHS to the LHS, you might get:<br>
+     * (select x x1, sum(x1) from y)<br>
+     * Now x1 is invalid.
+     */
+    public static boolean aliasAppearsInFuncArgs(QueryModel model, CharSequence alias, ArrayDeque<ExpressionNode> sqlNodeStack) {
+        if (model == null) {
+            return false;
+        }
+        boolean appearsInArgs = false;
+        ObjList<QueryColumn> modelColumns = model.getColumns();
+        for (int i = 0, n = modelColumns.size(); i < n; i++) {
+            QueryColumn col = modelColumns.get(i);
+            switch (col.getAst().type) {
```

Comment:
There used to be a `case LITERAL` check in this switch. We should keep it to handle scenarios like the following:
```sql
select X1 from (select x X1 from y) group by X1;
```

Sorry if I mislead you with this comment (https://github.com/questdb/questdb/pull/4190#discussion_r1480188400). The suggestion was to merge `appearsInCols` and `appearsInArgs` variables into a single one, so that both checks for function/operator and column update it:
```java
                case LITERAL:
                    appearsInArgs |= Chars.equalsIgnoreCase(col.getAst().token,alias);
                    break;
```

Please let me know if I'm missing something.